In [94]:
import pandas as pd
import numpy as np

In [95]:
RetailStore=pd.read_csv('retail_store_sales.csv')
RetailStore

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False
...,...,...,...,...,...,...,...,...,...,...,...
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,2023-09-03,NaN
12571,TXN_4009414,CUST_03,Beverages,Item_2_BEV,6.5,9.0,58.5,Cash,Online,2022-08-12,False
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,2024-08-24,NaN
12573,TXN_5167298,CUST_04,Furniture,Item_7_FUR,14.0,6.0,84.0,Cash,Online,2023-12-30,True


In [96]:
RetailStore.shape

(12575, 11)

In [97]:
RetailStore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [98]:
RetailStore.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [99]:
RetailStore.describe()

,Price Per Unit,Quantity,Total Spent
count,11966.000000,11971.000000,11971.000000
mean,23.365912,5.536380,129.652577
std,10.743519,2.857883,94.750697
min,5.000000,1.000000,5.000000
25%,14.000000,3.000000,51.000000
50%,23.000000,6.000000,108.500000
75%,33.500000,8.000000,192.000000
max,41.000000,10.000000,410.000000


In [100]:
RetailStore.duplicated().sum()


0

Sprint1:

*the dataset have 12575 rows and 11 coulmn
*The dataset have 8 columns (object ) and 3 columns(float)
*the dataset have a 7229 null values:
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Discount Applied    4199
*the data set has a 3 numrical columns:Price Per Unit,Quantity and Total Spent
*The describe for Price Per Unit have a min=5 ,max=41,median(seconde quartile)=23 and mean=23.365912 so don't have outliers (first research) 
*The describe for Quantity have a min=1 ,max=10,median(seconde quartile)=6 and mean=5.536380.365912 so don't have outliers (first research) 
*The describe for Total Spent have a min=5 ,max=410,median(seconde quartile)=108.5 and mean=129.652577 so don't have outliers (first research) 


In [101]:
RetailStore

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False
...,...,...,...,...,...,...,...,...,...,...,...
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,2023-09-03,NaN
12571,TXN_4009414,CUST_03,Beverages,Item_2_BEV,6.5,9.0,58.5,Cash,Online,2022-08-12,False
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,2024-08-24,NaN
12573,TXN_5167298,CUST_04,Furniture,Item_7_FUR,14.0,6.0,84.0,Cash,Online,2023-12-30,True


In [102]:
RetailStore['Transaction ID'].nunique()

12575

In [103]:
RetailStore['Customer ID'].value_counts()


Customer ID
CUST_05    544
CUST_24    543
CUST_13    534
CUST_08    533
CUST_09    519
CUST_15    519
CUST_16    515
CUST_23    513
CUST_20    507
CUST_18    507
CUST_01    507
CUST_11    503
CUST_22    501
CUST_10    501
CUST_21    498
CUST_12    498
CUST_07    491
CUST_02    488
CUST_17    487
CUST_19    487
CUST_14    484
CUST_06    481
CUST_25    476
CUST_04    474
CUST_03    465
Name: count, dtype: int64

In [104]:
RetailStore['Customer ID'].nunique()


25

In [105]:
RetailStore['Category'].value_counts()


Category
Furniture                             1591
Electric household essentials         1591
Food                                  1588
Milk Products                         1584
Butchers                              1568
Beverages                             1567
Computers and electric accessories    1558
Patisserie                            1528
Name: count, dtype: int64

In [106]:
RetailStore['Category'].nunique()

8

In [107]:
RetailStore['Item'].nunique()


200

In [108]:
RetailStore['Payment Method'].value_counts()


Payment Method
Cash              4310
Digital Wallet    4144
Credit Card       4121
Name: count, dtype: int64

In [109]:
RetailStore['Location'].nunique()
RetailStore['Location'].value_counts()


Location
Online      6354
In-store    6221
Name: count, dtype: int64

In [110]:
RetailStore['Discount Applied'].nunique()
RetailStore['Discount Applied'].value_counts()


Discount Applied
True     4219
False    4157
Name: count, dtype: int64

**spring2:**

-i have a 12575 unique Transaction ID , 25 Customer ID

**i have 8 categories:*

- Furniture ,Electric household essentials ,Food ,Milk Products ,Butchers, Beverages and Computers, electric accessories and Patisserie                

i have 200 items

**have a 3 payment Methode:**

- Cash              
- Digital Wallet   
- Credit Card      

**you can buy in two locations :**

- Online      
- In-store    

- The customer maybe gain a discount

- if the price is null so the item absolutly null
- if the price per unit is null so and quantitiy ,total spent exist so can calculate price per unit
- through price per unit=Total spent/Quantity 
- if i have a category and can calculate the price so can know what the item is null
- if i don't have a Quantity so the Total spent should be null
- if i don't know he total spent so i cann't apply discount or not

In [111]:
RetailStore[RetailStore['Discount Applied'].isnull()]
#if the price is null so the item absolutly null
#if the price per unit is null so and quantitiy ,total spent exist so can calculate price per unit
# through price per unit=Total spent/Quantity 
# if i have a category and can calculate the price so can know what the item is null
#if i don't have a Quantity so the Total spent should be null
#if i don't know he total spent so i cann't apply discount or not

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
5,TXN_7482416,CUST_09,Patisserie,NaN,NaN,10.0,200.0,Credit Card,Online,2023-11-30,NaN
14,TXN_2490363,CUST_09,Milk Products,Item_16_MILK,27.5,2.0,55.0,Digital Wallet,Online,2022-05-22,NaN
15,TXN_1809665,CUST_14,Beverages,NaN,24.5,NaN,NaN,Credit Card,In-store,2022-05-11,NaN
17,TXN_9634894,CUST_15,Milk Products,NaN,NaN,10.0,275.0,Digital Wallet,Online,2022-04-17,NaN
...,...,...,...,...,...,...,...,...,...,...,...
12562,TXN_7422454,CUST_07,Butchers,NaN,33.5,NaN,NaN,Cash,Online,2023-04-15,NaN
12568,TXN_1096134,CUST_06,Food,Item_4_FOOD,9.5,8.0,76.0,Digital Wallet,In-store,2022-06-11,NaN
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,2023-09-03,NaN
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,2024-08-24,NaN


In [112]:
RetailStore_copy=RetailStore.copy()

In [113]:
RetailStore_copy['Item'].unique()

array(['Item_10_PAT', 'Item_17_MILK', 'Item_12_BUT', 'Item_16_BEV',
       'Item_6_FOOD', nan, 'Item_1_FOOD', 'Item_16_FUR', 'Item_22_BUT',
       'Item_3_BUT', 'Item_2_FOOD', 'Item_24_PAT', 'Item_16_MILK',
       'Item_17_PAT', 'Item_13_EHE', 'Item_7_BEV', 'Item_4_EHE',
       'Item_10_FOOD', 'Item_14_FUR', 'Item_20_BUT', 'Item_25_FUR',
       'Item_14_FOOD', 'Item_22_PAT', 'Item_11_FOOD', 'Item_6_PAT',
       'Item_21_EHE', 'Item_25_BEV', 'Item_23_FOOD', 'Item_10_FUR',
       'Item_11_BEV', 'Item_23_BUT', 'Item_22_BEV', 'Item_10_EHE',
       'Item_24_BUT', 'Item_8_BEV', 'Item_3_FOOD', 'Item_12_FOOD',
       'Item_16_CEA', 'Item_11_PAT', 'Item_16_BUT', 'Item_5_CEA',
       'Item_19_MILK', 'Item_23_FUR', 'Item_7_FUR', 'Item_15_CEA',
       'Item_6_MILK', 'Item_24_CEA', 'Item_22_CEA', 'Item_22_FOOD',
       'Item_2_BUT', 'Item_14_PAT', 'Item_12_PAT', 'Item_18_FOOD',
       'Item_1_PAT', 'Item_4_BEV', 'Item_22_FUR', 'Item_7_PAT',
       'Item_20_CEA', 'Item_20_FOOD', 'Item_11_FUR', 'Item

In [114]:
RetailStore_copy.isnull().sum()

Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [115]:
mask = (
    RetailStore_copy['Price Per Unit'].isnull() &
    RetailStore_copy['Quantity'].notnull() &
    RetailStore_copy['Total Spent'].notnull()
)

RetailStore_copy.loc[mask, 'Price Per Unit'] = (
    RetailStore_copy.loc[mask, 'Total Spent'] /
    RetailStore_copy.loc[mask, 'Quantity']
)

In [116]:
RetailStore_copy.isnull().sum()


Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit         0
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [117]:
item_price_map = (
    RetailStore_copy
    .dropna(subset=['Item', 'Category', 'Price Per Unit'])
    .groupby(['Category', 'Price Per Unit'])['Item']
    .nunique()
)

item_price_map[item_price_map > 1]

Series([], Name: Item, dtype: int64)

In [118]:
item_lookup = (
    RetailStore_copy[RetailStore_copy['Item'].notnull()]
    [['Category', 'Price Per Unit', 'Item']]
    .drop_duplicates()
)

RetailStore_copy = RetailStore_copy.merge(
    item_lookup,
    on=['Category', 'Price Per Unit'],
    how='left',
    suffixes=('', '_new')
)

RetailStore_copy['Item'] = (
    RetailStore_copy['Item']
    .fillna(RetailStore_copy['Item_new'])
)

RetailStore_copy.drop(columns=['Item_new'], inplace=True)

In [119]:
RetailStore_copy[
    RetailStore_copy['Item'].isnull()
][['Category','Price Per Unit']].drop_duplicates()

,Category,Price Per Unit


In [120]:
check = (
    RetailStore_copy
    .groupby(['Category', 'Price Per Unit'])['Item']
    .nunique()
    .reset_index(name='Unique_Items')
)

check[check['Unique_Items'] > 1]

,Category,Price Per Unit,Unique_Items


In [121]:
RetailStore_copy.isnull().sum()


Transaction ID         0
Customer ID            0
Category               0
Item                   0
Price Per Unit         0
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64

In [122]:
RetailStore_copy[(RetailStore_copy['Quantity'].isnull())&(RetailStore_copy['Total Spent'].isnull())]

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
7,TXN_1372952,CUST_21,Furniture,Item_20_FUR,33.5,NaN,NaN,Digital Wallet,In-store,2024-04-02,True
15,TXN_1809665,CUST_14,Beverages,Item_14_BEV,24.5,NaN,NaN,Credit Card,In-store,2022-05-11,NaN
19,TXN_4206593,CUST_01,Furniture,Item_21_FUR,35.0,NaN,NaN,Digital Wallet,Online,2025-01-13,False
25,TXN_3481599,CUST_05,Furniture,Item_24_FUR,39.5,NaN,NaN,Cash,Online,2022-09-08,False
34,TXN_1621497,CUST_06,Patisserie,Item_13_PAT,23.0,NaN,NaN,Cash,In-store,2023-02-18,NaN
...,...,...,...,...,...,...,...,...,...,...,...
12527,TXN_1069238,CUST_23,Food,Item_1_FOOD,5.0,NaN,NaN,Digital Wallet,In-store,2022-08-13,False
12552,TXN_4823896,CUST_05,Milk Products,Item_3_MILK,8.0,NaN,NaN,Cash,In-store,2022-07-21,False
12556,TXN_4397672,CUST_04,Beverages,Item_25_BEV,41.0,NaN,NaN,Credit Card,Online,2024-11-28,True
12562,TXN_7422454,CUST_07,Butchers,Item_20_BUT,33.5,NaN,NaN,Cash,Online,2023-04-15,NaN


In [123]:
RetailStore_copy['Discount Applied'].value_counts(dropna=False)

Discount Applied
True     4219
NaN      4199
False    4157
Name: count, dtype: int64

In [124]:
RetailStore_copy['Discount Applied'] = RetailStore_copy['Discount Applied'].fillna(False)

In [125]:
RetailStore_copy['Discount Applied'].value_counts(dropna=False)


Discount Applied
False    8356
True     4219
Name: count, dtype: int64

In [126]:
complete_rows = RetailStore_copy.dropna(
    subset=['Price Per Unit', 'Quantity', 'Total Spent']
).copy()

complete_rows['Expected Total'] = (
    complete_rows['Price Per Unit'] *
    complete_rows['Quantity']
)

inconsistent_rows = complete_rows[
    complete_rows['Expected Total'] != complete_rows['Total Spent']
]

print("Complete rows:", len(complete_rows))
print("Inconsistent rows:", len(inconsistent_rows))
print("Inconsistency percentage:",
      len(inconsistent_rows) / len(complete_rows) * 100)

Complete rows: 11971
Inconsistent rows: 0
Inconsistency percentage: 0.0


In [127]:
RetailStore_copy

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,False
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False
...,...,...,...,...,...,...,...,...,...,...,...
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,2023-09-03,False
12571,TXN_4009414,CUST_03,Beverages,Item_2_BEV,6.5,9.0,58.5,Cash,Online,2022-08-12,False
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,2024-08-24,False
12573,TXN_5167298,CUST_04,Furniture,Item_7_FUR,14.0,6.0,84.0,Cash,Online,2023-12-30,True


In [128]:
RetailStore_copy['Transaction Date'] = pd.to_datetime(
    RetailStore_copy['Transaction Date'],
    errors='coerce'
)

RetailStore_copy['Year'] = RetailStore_copy['Transaction Date'].dt.year
RetailStore_copy['Month'] = RetailStore_copy['Transaction Date'].dt.month
RetailStore_copy['Day'] = RetailStore_copy['Transaction Date'].dt.day


In [129]:
RetailStore_copy

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Year,Month,Day
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True,2024,4,8
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True,2023,7,23
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False,2022,10,5
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,False,2022,5,7
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False,2022,10,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,2023-09-03,False,2023,9,3
12571,TXN_4009414,CUST_03,Beverages,Item_2_BEV,6.5,9.0,58.5,Cash,Online,2022-08-12,False,2022,8,12
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,2024-08-24,False,2024,8,24
12573,TXN_5167298,CUST_04,Furniture,Item_7_FUR,14.0,6.0,84.0,Cash,Online,2023-12-30,True,2023,12,30


In [130]:
RetailStore_copy.drop(columns='Transaction Date')

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Discount Applied,Year,Month,Day
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,True,2024,4,8
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,True,2023,7,23
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,False,2022,10,5
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,False,2022,5,7
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,False,2022,10,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
12570,TXN_9347481,CUST_18,Patisserie,Item_23_PAT,38.0,4.0,152.0,Credit Card,In-store,False,2023,9,3
12571,TXN_4009414,CUST_03,Beverages,Item_2_BEV,6.5,9.0,58.5,Cash,Online,False,2022,8,12
12572,TXN_5306010,CUST_11,Butchers,Item_7_BUT,14.0,10.0,140.0,Cash,Online,False,2024,8,24
12573,TXN_5167298,CUST_04,Furniture,Item_7_FUR,14.0,6.0,84.0,Cash,Online,True,2023,12,30


In [131]:
transactions_per_month = (
    RetailStore_copy
    .groupby(['Year', 'Month'])['Transaction ID']
    .nunique()
    .reset_index(name='Number_of_Transactions')
    .sort_values(['Year', 'Month'])
)

transactions_per_month

#The number of transactions in all month Converging values 

,Year,Month,Number_of_Transactions
0,2022,1,390
1,2022,2,331
2,2022,3,353
3,2022,4,346
4,2022,5,330
5,2022,6,338
6,2022,7,371
7,2022,8,347
8,2022,9,348
9,2022,10,317


In [132]:
transactions_per_month = (
    RetailStore_copy
    .groupby('Month')['Transaction ID']
    .nunique()
    .reset_index(name='Number_of_Transactions')
)

transactions_per_month
#The number of transaction in all month from years have a converging values 
#in month 1(january) has a most transactions because in 2025 have data just for january,data is not completed or finish in this month


,Month,Number_of_Transactions
0,1,1361
1,2,965
2,3,1019
3,4,996
4,5,1033
5,6,1038
6,7,1089
7,8,1039
8,9,1021
9,10,978


In [133]:
customers_per_month = (
    RetailStore_copy
    .groupby(['Year', 'Month'])['Customer ID']
    .nunique()
    .reset_index(name='Number_of_Customers')
    .sort_values(['Year', 'Month'])
)

customers_per_month

,Year,Month,Number_of_Customers
0,2022,1,25
1,2022,2,25
2,2022,3,25
3,2022,4,25
4,2022,5,25
5,2022,6,25
6,2022,7,25
7,2022,8,25
8,2022,9,25
9,2022,10,25


In [134]:
customers_per_month = (
    RetailStore_copy
    .groupby(['Month'])['Customer ID']
    .nunique()
    .reset_index(name='Number_of_Customers')
    .sort_values(['Month'])
)

customers_per_month

,Month,Number_of_Customers
0,1,25
1,2,25
2,3,25
3,4,25
4,5,25
5,6,25
6,7,25
7,8,25
8,9,25
9,10,25


In [135]:
RetailStore_copy['Customer ID'].nunique()

25

In [136]:
customer_sets = (
    RetailStore_copy
    .groupby(['Year', 'Month'])['Customer ID']
    .agg(lambda x: frozenset(x))
)

customer_sets.nunique()
#number of customers in all month in every year is same and all customers make a transactions monthly

1

In [137]:
top_customer_each_month = (
    RetailStore_copy
    .groupby(['Year', 'Month', 'Customer ID'])['Transaction ID']
    .nunique()
    .reset_index(name='Number_of_Transactions')
    .assign(
        Max_Transactions=lambda df: df.groupby(['Year', 'Month'])[
            'Number_of_Transactions'
        ].transform('max')
    )
    .query('Number_of_Transactions == Max_Transactions')
    .drop(columns='Max_Transactions')
    .sort_values(['Year', 'Month', 'Customer ID'])
)

top_customer_each_month

,Year,Month,Customer ID,Number_of_Transactions
16,2022,1,CUST_17,21
33,2022,2,CUST_09,19
48,2022,2,CUST_24,19
68,2022,3,CUST_19,22
76,2022,4,CUST_02,21
101,2022,5,CUST_02,18
147,2022,6,CUST_23,24
173,2022,7,CUST_24,28
190,2022,8,CUST_16,26
219,2022,9,CUST_20,20


In [138]:
top_customer_each_month = (
    RetailStore_copy
    .groupby(['Year', 'Month', 'Customer ID'])['Transaction ID']
    .nunique()
    .reset_index(name='Number_of_Transactions')
    .assign(
        Max_Transactions=lambda df: df.groupby(['Year', 'Month'])[
            'Number_of_Transactions'
        ].transform('min')
    )
    .query('Number_of_Transactions == Max_Transactions')
    .drop(columns='Max_Transactions')
    .sort_values(['Year', 'Month', 'Customer ID'])
)

top_customer_each_month

,Year,Month,Customer ID,Number_of_Transactions
1,2022,1,CUST_02,8
44,2022,2,CUST_20,5
58,2022,3,CUST_09,8
72,2022,3,CUST_23,8
90,2022,4,CUST_16,8
104,2022,5,CUST_05,9
109,2022,5,CUST_10,9
120,2022,5,CUST_21,9
143,2022,6,CUST_19,6
167,2022,7,CUST_18,8


In [139]:
top_item_each_month = (
    RetailStore_copy
    .groupby(['Year', 'Month', 'Item'])['Quantity']
    .sum()
    .reset_index(name='Total_Quantity_Sold')
    .assign(
        Max_Quantity=lambda df: df.groupby(['Year', 'Month'])[
            'Total_Quantity_Sold'
        ].transform('max')
    )
    .query('Total_Quantity_Sold == Max_Quantity')
    .drop(columns='Max_Quantity')
    .sort_values(['Year', 'Month', 'Item'])
)

top_item_each_month

,Year,Month,Item,Total_Quantity_Sold
13,2022,1,Item_12_BUT,61.0
216,2022,2,Item_1_FUR,38.0
364,2022,3,Item_1_MILK,42.0
386,2022,3,Item_23_EHE,42.0
510,2022,4,Item_1_MILK,39.0
629,2022,5,Item_15_EHE,41.0
844,2022,6,Item_2_BEV,43.0
864,2022,6,Item_5_FUR,43.0
990,2022,7,Item_25_EHE,48.0
1142,2022,8,Item_2_BEV,52.0


In [140]:
top_item_each_customer = (
    RetailStore_copy
    .groupby(['Customer ID', 'Item'])['Quantity']
    .sum()
    .reset_index(name='Total_Quantity_Bought')
    .assign(
        Max_Quantity=lambda df: df.groupby('Customer ID')[
            'Total_Quantity_Bought'
        ].transform('max')
    )
    .query('Total_Quantity_Bought == Max_Quantity')
    .drop(columns='Max_Quantity')
    .sort_values(['Customer ID', 'Item'])
)

top_item_each_customer

,Customer ID,Item,Total_Quantity_Bought
111,CUST_01,Item_2_BEV,55.0
181,CUST_02,Item_12_BEV,57.0
479,CUST_03,Item_6_FOOD,53.0
614,CUST_04,Item_2_BEV,66.0
779,CUST_05,Item_25_BUT,61.0
968,CUST_06,Item_3_MILK,47.0
1135,CUST_07,Item_5_FOOD,47.0
1199,CUST_08,Item_15_FUR,54.0
1349,CUST_09,Item_13_MILK,57.0
1625,CUST_10,Item_5_CEA,51.0


In [141]:
top_item_each_customer = (
    RetailStore_copy
    .groupby(['Customer ID', 'Item'])['Quantity']
    .size()
    .reset_index(name='Total_Quantity_Bought')
    .assign(
        Max_Quantity=lambda df: df.groupby('Customer ID')[
            'Total_Quantity_Bought'
        ].transform('max')
    )
    .query('Total_Quantity_Bought == Max_Quantity')
    .drop(columns='Max_Quantity')
    .sort_values(['Customer ID', 'Item'])
)

top_item_each_customer

,Customer ID,Item,Total_Quantity_Bought
135,CUST_01,Item_5_EHE,9
182,CUST_02,Item_12_BUT,10
428,CUST_03,Item_23_BUT,9
444,CUST_03,Item_25_FOOD,9
562,CUST_04,Item_19_CEA,10
779,CUST_05,Item_25_BUT,9
957,CUST_06,Item_2_BEV,8
958,CUST_06,Item_2_BUT,8
1102,CUST_07,Item_24_FUR,8
1126,CUST_07,Item_4_BEV,8


In [142]:
revenue_by_item = (
    RetailStore_copy
    .groupby('Item')['Total Spent']
    .sum()
    .reset_index(name='Total_Revenue')
    .sort_values('Total_Revenue', ascending=False)
)

revenue_by_item.head(10)

,Item,Total_Revenue
133,Item_25_FUR,25625.0
131,Item_25_EHE,25502.0
129,Item_25_BUT,22427.0
125,Item_24_FUR,22080.5
132,Item_25_FOOD,21771.0
105,Item_22_BUT,21462.0
113,Item_23_BUT,20596.0
78,Item_19_MILK,20448.0
115,Item_23_EHE,20178.0
119,Item_23_PAT,19684.0


In [143]:
revenue_by_item.tail(10)

,Item,Total_Revenue
29,Item_13_FUR,920.0
198,Item_9_MILK,850.0
24,Item_13_BEV,828.0
103,Item_21_PAT,805.0
140,Item_2_FOOD,793.0
84,Item_1_FOOD,715.0
82,Item_1_CEA,645.0
154,Item_4_CEA,636.5
160,Item_5_BEV,517.0
147,Item_3_EHE,224.0


In [144]:
quantity_by_item = (
    RetailStore_copy
    .groupby('Item')['Quantity']
    .sum()
    .reset_index(name='Total_Quantity')
    .sort_values('Total_Quantity', ascending=False)
)

quantity_by_item.head(10)

,Item,Total_Quantity
136,Item_2_BEV,693.0
54,Item_16_MILK,664.0
78,Item_19_MILK,639.0
36,Item_14_FOOD,628.0
28,Item_13_FOOD,628.0
165,Item_5_FUR,625.0
133,Item_25_FUR,625.0
86,Item_1_MILK,622.0
131,Item_25_EHE,622.0
74,Item_19_CEA,603.0


In [145]:
quantity_by_item.tail(10)

,Item,Total_Quantity
122,Item_24_CEA,54.0
6,Item_10_MILK,54.0
198,Item_9_MILK,50.0
116,Item_23_FOOD,49.0
109,Item_22_FUR,48.0
160,Item_5_BEV,47.0
29,Item_13_FUR,40.0
24,Item_13_BEV,36.0
147,Item_3_EHE,28.0
103,Item_21_PAT,23.0


In [146]:
revenue_by_category = (
    RetailStore_copy
    .groupby('Category')['Total Spent']
    .sum()
    .reset_index(name='Total_Revenue')
    .sort_values('Total_Revenue', ascending=False)
)

revenue_by_category

,Category,Total_Revenue
1,Butchers,208118.0
3,Electric household essentials,203813.5
0,Beverages,197047.5
5,Furniture,195310.0
4,Food,194812.0
2,Computers and electric accessories,190692.5
7,Patisserie,182165.5
6,Milk Products,180112.0


In [147]:
quantity_by_category = (
    RetailStore_copy
    .groupby('Category')['Quantity']
    .sum()
    .reset_index(name='Total_Quantity')
    .sort_values('Total_Quantity', ascending=False)
)

quantity_by_category

,Category,Total_Quantity
5,Furniture,8462.0
4,Food,8387.0
0,Beverages,8358.0
6,Milk Products,8339.0
3,Electric household essentials,8309.0
2,Computers and electric accessories,8272.0
1,Butchers,8206.0
7,Patisserie,7943.0


In [148]:
transactions_by_payment = (
    RetailStore_copy
    .groupby('Payment Method')['Transaction ID']
    .nunique()
    .reset_index(name='Transactions')
    .sort_values('Transactions', ascending=False)

)

transactions_by_payment

,Payment Method,Transactions
0,Cash,4310
2,Digital Wallet,4144
1,Credit Card,4121


In [149]:
revenue_by_payment = (
    RetailStore_copy
    .groupby('Payment Method')['Total Spent']
    .sum()
    .reset_index(name='Total_Revenue')
    .sort_values('Total_Revenue', ascending=False)
)

revenue_by_payment

,Payment Method,Total_Revenue
0,Cash,537710.0
2,Digital Wallet,507279.0
1,Credit Card,507082.0


In [150]:
avg_transaction_payment = (
    RetailStore_copy
    .groupby('Payment Method')['Total Spent']
    .mean()
    .reset_index(name='Average_Transaction_Value')
)

avg_transaction_payment

,Payment Method,Average_Transaction_Value
0,Cash,131.052888
1,Credit Card,129.127069
2,Digital Wallet,128.718346


In [151]:
transactions_discount = (
    RetailStore_copy
    .groupby('Discount Applied')['Transaction ID']
    .nunique()
    .reset_index(name='Transactions')
        .sort_values('Transactions', ascending=False)

)

transactions_discount

,Discount Applied,Transactions
0,False,8356
1,True,4219


In [152]:
revenue_discount = (
    RetailStore_copy
    .groupby('Discount Applied')['Total Spent']
    .sum()
    .reset_index(name='Total_Revenue')
        .sort_values('Total_Revenue', ascending=False)

)

revenue_discount

,Discount Applied,Total_Revenue
0,False,1027627.5
1,True,524443.5


In [153]:
avg_transaction_discount = (
    RetailStore_copy
    .groupby('Discount Applied')['Total Spent']
    .mean()
    .reset_index(name='Average_Transaction_Value')
)

avg_transaction_discount

,Discount Applied,Average_Transaction_Value
0,False,129.228810
1,True,130.491043


In [154]:
avg_quantity_discount = (
    RetailStore_copy
    .groupby('Discount Applied')['Quantity']
    .mean()
    .reset_index(name='Average_Quantity')
)

avg_quantity_discount

,Discount Applied,Average_Quantity
0,False,5.537852
1,True,5.533466


In [155]:
transactions_location = (
    RetailStore_copy
    .groupby('Location')['Transaction ID']
    .nunique()
    .reset_index(name='Transactions')
)

transactions_location

,Location,Transactions
0,In-store,6221
1,Online,6354


In [156]:
revenue_location = (
    RetailStore_copy
    .groupby('Location')['Total Spent']
    .sum()
    .reset_index(name='Total_Revenue')
)

revenue_location

,Location,Total_Revenue
0,In-store,760670.0
1,Online,791401.0


In [157]:
avg_transaction_location = (
    RetailStore_copy
    .groupby('Location')['Total Spent']
    .mean()
    .reset_index(name='Average_Transaction_Value')
)

avg_transaction_location

,Location,Average_Transaction_Value
0,In-store,128.861596
1,Online,130.422050


In [158]:
avg_quantity_location = (
    RetailStore_copy
    .groupby('Location')['Quantity']
    .mean()
    .reset_index(name='Average_Quantity')
)

avg_quantity_location

,Location,Average_Quantity
0,In-store,5.513298
1,Online,5.558833


In [159]:
highest_price = (
    RetailStore_copy
    .groupby('Item')['Price Per Unit']
    .max()
    .reset_index(name='Highest_Price')
    .sort_values('Highest_Price', ascending=False)
)

highest_price.head(10)

,Item,Highest_Price
129,Item_25_BUT,41.0
135,Item_25_PAT,41.0
134,Item_25_MILK,41.0
133,Item_25_FUR,41.0
132,Item_25_FOOD,41.0
131,Item_25_EHE,41.0
130,Item_25_CEA,41.0
128,Item_25_BEV,41.0
120,Item_24_BEV,39.5
127,Item_24_PAT,39.5


In [160]:
highest_quantity = (
    RetailStore_copy
    .groupby('Item')['Quantity']
    .sum()
    .reset_index(name='Total_Quantity')
    .sort_values('Total_Quantity', ascending=False)
)

highest_quantity.head(10)

,Item,Total_Quantity
136,Item_2_BEV,693.0
54,Item_16_MILK,664.0
78,Item_19_MILK,639.0
36,Item_14_FOOD,628.0
28,Item_13_FOOD,628.0
165,Item_5_FUR,625.0
133,Item_25_FUR,625.0
86,Item_1_MILK,622.0
131,Item_25_EHE,622.0
74,Item_19_CEA,603.0


In [161]:
RetailStore_copy.to_csv("Retail.csv", index=False)